### ISSUE TITLE
Yaml-cpp bug in debug mode

### ISSUE THREAD

**Issue by breznak** [htm-community/htm.core#218]
Unit tests don't run on Linux in Debug mode, due to upstream bug in yaml-cpp, reported here 
https://github.com/jbeder/yaml-cpp/issues/658 

The test in question: 
```
CppRegionTest.testCppLinkingFanIn
Neoprávněný přístup do paměti (SIGSEGV)
```

EDIT: 
- [ ] re-enable `CppRegionTest` even for Debug when resolved

**Comment by breznak** [comment]
Yaml-cpp does seem to be rather slow on development and merging PRs, is there an alternative? (I've searched and it's basically the only lib for YAML c++ found, other is some ThorsSerializer, that's all) 

Possible steps: 
- alternative lib
- build deps (and tests) without debug, only src/ is passed the debug flag? 
- wait for upstream fix

**Comment by breznak** [comment]
Moving the discussion back to the issue from https://github.com/htm-community/nupic.cpp/pull/282#issuecomment-468043819 

```
 Ok, here is what I am finding:

    For C++, yaml-cpp is about the only game in town.
    Nobody else seems to be having a crash like ours. That makes me suspicious that there is something we are not doing right. There is one thing that most people do not do is compile it with -fPIC option. But, we need that because this code needs to go into our Python extension library which is basically a shared library.
    We call yaml-cpp in just one place:

      const YAML::Node doc = YAML::Load(parameter_string);

    The returned doc is a tree of parsed nodes. The crash is inside the Load( ) function. Nothing strange about how it is called.
    As an alternative there is libyaml which is a yaml parser written in C which is quite popular. We could add a C++ wrapper. This is a SAX parser which means it performs callbacks for each field during the parse rather than returning a nested tree structure like yaml-cpp. I could adapt our code to use this without too much problem since what we do is iterate yaml-cpp's structure and build another one anyway. It might actually be faster to use callbacks but since this is only used at startup, speed is not really an issue.

At this point, I am ready to try the libyaml parser. I will do this in parallel so that I don't take out yaml-cpp until I am sure libyaml will work for us.

Sorry to be so verbose on this topic but I wanted to document the decisions.
```

Thanks @dkeeney ! 

Ok, so I can reproduce this and have some findings: 

- Release mode: OK
- Debug: segfault on gcc 7.3, gcc 8.2 (on Ubuntu 18.04 LTS) 
- Debug on clang 6.0 is OK! 
- it matters how (which compiler) yaml-cpp (external dep) is built
  - does not matter how "our lib" is built. So just `rm -rf build/scripts/*`, or `make clean` and `cmake` is not enough!
  - one must delete & rebuild dependencies (yaml-cpp), so `cd $REPO/; git clean -xdf` !caution, will remove all unchecked files from your repo!! 
- I tried building yaml-cpp in release only, but does not help 


For next steps, I suggest the following: 
- please push the patch to build yaml-cpp in Release always
- please update the upstream issue with the info you've found here, link this issue, the PR. 
  - try to create a minimal reproducible test, submit PR upstream to yaml-cpp 
  - I'd set this to "low" priority and wait for some upstream reaction
  - do not convert to libyaml (c) yet, this issue does not block us right now

**Comment by dkeeney** [comment]
> one must delete & rebuild dependencies (yaml-cpp), so

What I do is `cd REPO/; rm -rf build`
This gets rid of the ThirdParty folder without having to loose any changed source files.
Then rebuild with with debug mode:
```
 mkdir -p build/scripts; cd build/scripts
 cmake -DCMAKE_BUILD_TYPE=Debug ../..
``
I tried forcing yaml-cpp to build as release mode always with everything else in debug mode and it still fails in the same way.

>  and wait for some upstream reaction

Not likely.  The comments I have seen on the web is that there is not much going on to maintain yaml-cpp.  Even Google offered to assist in fixing a problem they had with no activity.

>  do not convert to libyaml (c) yet, this issue does not block us right now

You are probably right, this is rather low priority at this point.  I have a header-only C++ wrapper about half finished that will make libyaml look very much like yaml-cpp so that I would not need to change anything in NetworkAPI to use it.  It is probably 2-3 days more work.

I will mothball what I have and go back to working on TMRegion.
I have the MSVC to debug with.  Actually I like the debugger in MSVC better than anything I have yet seen on Linux anyway.  But that does not help you folks.  I can provide a valid Windows7 license if you would want setup a VurtualBox window on your system.

**Comment by breznak** [comment]
There's a new cpp YAML parsing library, actively developed. Might be worth looking at: 
https://github.com/biojppm/rapidyaml

**Comment by dkeeney** [comment]
> There's a new cpp YAML parsing library

Oh, this is perfect.  Well spotted.

**Comment by breznak** [comment]
Fixed in #715 by using a different yaml parser backend - libyaml.
